src/CoreWCF.Primitives/src/CoreWCF/Security/SecuritySessionServerSettings.cs

```diff
@@ -1956,6 +1967,562 @@ public Task OpenAsync(CancellationToken token)
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.ServerReceivedCloseMessageStateIsCreated, GetType().ToString())));
+                }
+
+                bool setInputSessionCloseHandle = false;
+                bool cleanupContext = true;
+                try
+                {
+                    lock (LocalLock)
+                    {
+                        _receivedClose = true;
+                        if (!_isInputClosed)
+                        {
+                            _isInputClosed = true;
+                            setInputSessionCloseHandle = true;
+
+                            if (!_isOutputClosed)
+                            {
+                                _closeRequestContext = requestContext;
+                                // CreateCloseResponse closes the message passed in
+                                _closeResponseMessage = CreateCloseResponse(message, null, token);
+                                cleanupContext = false;
+                            }
+                        }
+                    }
+
+                    if (setInputSessionCloseHandle)
+                    {
+                       _inputSessionCloseHandle.Set();
+                    }
+                    if (cleanupContext)
+                    {
+                        requestContext.CloseAsync();
+                        cleanupContext = false;
+                    }
+                }
+                finally
+                {
+                    message.Close();
+                    if (cleanupContext)
+                    {
+                        requestContext.Abort();
+                    }
+                }
+            }
+
+            protected override void OnCloseResponseMessageReceived(RequestContext requestContext, Message message, SecurityProtocolCorrelationState correlationState, TimeSpan timeout)
+            {
+                bool cleanupContext = true;
+                try
+                {
+                    bool isCloseResponseExpected = false;
+                    bool setInputSessionCloseHandle = false;
+                    lock (LocalLock)
+                    {
+                        isCloseResponseExpected = _sentClose;
+                        if (isCloseResponseExpected && !_isInputClosed)
+                        {
+                            _isInputClosed = true;
+                            setInputSessionCloseHandle = true;
+                        }
+                    }
+                    if (!isCloseResponseExpected)
+                    {
+                        _duplexCommObj.Fault(new ProtocolException(SR.Format(SR.UnexpectedSecuritySessionCloseResponse)));
+                        return;
+                    }
+                    if (setInputSessionCloseHandle)
+                    {
+                        _inputSessionCloseHandle.Set();
+                    }
```

Comment:
You are inconsistent in restoring the code which sets the handle. This will result in deadlocks if you don't make sure the same places do the same thing with the handle.

Reply:
this is kept for NetTCP (Duplex) mode...